### PR TITLE
[SPARK-14214][SQL] Update state to provide get/put interface

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

The goal is to make the state store more flexible by giving it a more key-value store like interface of gets and puts. This is to allow streaming aggregates physical plan to easily do only-gets instead of only-updates that the current is designed for.
## How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by tdas** [comment]
@marmbrus @zsxwing

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala [hunk 1]
```diff
@@ -54,17 +54,10 @@ class StateStoreRDD[T: ClassTag, U: ClassTag](
...
-      assert(store.hasCommitted)
-      outputIter
-    } {
-      if (store != null) store.cancel()
-    }
+    val storeId = StateStoreId(checkpointLocation, operatorId, partition.index)
+    store = StateStore.get(
+      storeId, keySchema, valueSchema, storeVersion, storeConf, confBroadcast.value.value)
+    val inputIter = dataRDD.iterator(partition, ctxt)
+    storeUpdateFunction(store, inputIter)
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreRDD.scala]: The finally was removed to allow commits to be done lazily.

**Comment by tdas** [comment]
Not sure about what this mima test failure is about. Thats anyways not on the latest commit, so I will wait for the tests on the latest commit to completed (build # 2703).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala [hunk 2]
```diff
@@ -148,7 +146,7 @@ private[state] class HDFSBackedStateStoreProvider(
 
     /** Commit all the updates that have been made to the store, and return the new version. */
     override def commit(): Long = {
-      verify(state == UPDATING, "Cannot commit again after already committed or cancelled")
+      verify(state == UPDATING, "Cannot commit after already committed or cancelled")
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala]: nit: cancelled -> aborted

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala [hunk 3]
```diff
@@ -94,15 +94,14 @@ private[state] class HDFSBackedStateStoreProvider(
...
-    override def update(key: UnsafeRow, updateFunc: Option[UnsafeRow] => UnsafeRow): Unit = {
-      verify(state == UPDATING, "Cannot update after already committed or cancelled")
-      val oldValueOption = Option(mapToUpdate.get(key))
-      val value = updateFunc(oldValueOption)
+    override def get(key: UnsafeRow): Option[UnsafeRow] = {
+      Option(mapToUpdate.get(key))
+    }
+
+    override def put(key: UnsafeRow, value: UnsafeRow): Unit = {
+      verify(state == UPDATING, "Cannot remove after already committed or cancelled")
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/HDFSBackedStateStoreProvider.scala]: nit: remove -> put

**Comment by zsxwing** [comment]
LGTM except one nit

**Comment by marmbrus** [comment]
Can we close this now?

**Comment by tdas** [comment]
closing this as #12048 already absorbed this change.

**Pull request closed without merging**